### DISCUSSION TITLE
Disjointed parts of a recognized object mask

### DISCUSSION THREAD

**Discussion by jadrian21** [ultralytics/ultralytics#10030] [Docs]
I've noticed that I can only plot the points of the first segment of a disjointed object mask recognition and I can't find the other parts of the disjointed mask to plot them as they appear not to be in the list results.mask.xy. Where are they stored? for example in one recognition
frame I have a person sitting on a bed, the only mask segment I can plot is to the left of the person, all the two other parts, between the
legs of the person and to their right are missing.

**Comment by glenn-jocher** [comment]
Hello! 🌟

It sounds like you're encountering an issue with visualizing all segments of disjointed object masks. For instance, when using `.masks` attribute from the model's results, it provides segmentation masks for detected objects, but it might not separate each disjointed segment for an individual object explicitly.

However, every detected mask segment, regardless of whether they belong to disjointed parts of an object, should typically be included within the `results.masks` attribute as separate entities. If you're missing segments, it could be a matter of how these are being visualized or interpreted.

Here's a quick example on accessing and visualizing mask segments:

```python
from ultralytics import YOLO
import matplotlib.pyplot as plt
import numpy as np

model = YOLO('yolov8n-seg.pt')  # Ensure you're using a segmentation model
results = model('image.jpg')

for i, mask in enumerate(results.masks):  # Iterate over each detected mask
    # Access mask tensor and convert it to numpy for visualization
    mask_np = mask.numpy()[0]
    plt.imshow(mask_np, cmap='gray', interpolation='none')
    plt.title(f'Mask {i}')
    plt.show()
```
This code should help in visualizing each detected mask, including disjointed parts, assuming they were recognized as separate entities by the model. If certain segments are indeed not detected, it might be a limitation or requirement for further tuning of the model based on your specific images. 

In some cases, improving detection on such specific scenarios could involve retraining the model with more representative data of the objects in varied configurations, ensuring that all potential disjointed parts are adequately represented in the training dataset.

Hope this helps, and feel free to share more details if you're still facing issues! 🚀

**Reply by DrukSonam** [reply]
With this code i get error: AttributeError: 'list' object has no attribute 'masks'. I am using yolov8

**Reply by glenn-jocher** [reply]
@DrukSonam hello! 👋

It seems like you're encountering an issue because `results` is a list of `Results` objects, not a single object. You'll need to iterate through this list to access the `masks` attribute for each `Results` object.

Here's an updated example:

```python
from ultralytics import YOLO
import matplotlib.pyplot as plt

model = YOLO('yolov8n-seg.pt')  # Ensure you're using a segmentation model
results = model('image.jpg')

for result in results:  # Iterate over each Results object
    for i, mask in enumerate(result.masks):  # Iterate over each detected mask
        mask_np = mask.numpy()[0]
        plt.imshow(mask_np, cmap='gray', interpolation='none')
        plt.title(f'Mask {i}')
        plt.show()
```

This should help you visualize each detected mask. If you need further assistance, feel free to ask! 😊
